### PR TITLE
Add logs to OpenAPI spec and client 

### PR DESCRIPTION
**Pull request by nyza99**
## What?

This PR adds support for logs API routes to OpenAPI spec for the Observatorium API and generated client.

## Why?

The current [OpenAPI ](https://github.com/observatorium/api/tree/main/client)is missing support for the [observatorium/api logs API](https://github.com/observatorium/api/blob/main/api/logs/v1/http.go#L90). This make writing client code (e.g. in obsctl) error-prone and cumbersome. In addition the logs api remains as such undocumented as per OpenAPI spec can be used to generate user-friendly documentation (See https://github.com/observatorium/observatorium/pull/483)

## How?

Added support for the following logs API routes in the OpenAPI spec:

```
POST /api/logs/v1/<tenant>/loki/api/v1/push
GET /api/logs/v1/<tenant>/loki/api/v1/labels
GET /api/logs/v1/<tenant>/loki/api/v1/label/<name>/values/
GET /api/logs/v1/<tenant>/loki/api/v1/series
GET /api/logs/v1/<tenant>/loki/api/v1/query_range
GET /api/logs/v1/<tenant>/loki/api/v1/query
GET /api/logs/v1/<tenant>/loki/api/v1/tail
```

## Testing?

For testing run `make test-interactive` 

This command spins up a full setup for Observatorium API with all dependent services. It is intended for short-lived manual testing on your local environment.

Run Curl commands to make API calls against Observatorium API to test manually. 

example:

`curl --insecure -X POST -v -H “Content-Type: application/json” -H “Accept: application/json” -H 'Authorization: Bearer <API TOKEN>  "https://127.0.0.1:<observatorium API port>/api/logs/v1/<tenant>/loki/api/v1/push” -d '{"streams": [{ "stream": { "foo": "bar1" }, "values": [[ "1570818238000000000", "fizzbuzz"]]}]}' `

`curl --insecure  -G -v -H "Content-Type: application/json" -H 'Authorization: Bearer <API TOKEN> ' "https://127.0.0.1:<observatorium API port>/api/logs/v1/<tenant>/loki/api/v1/query" --data-urlencode 'query=count_over_time({_id="test"}[1000m])'`

## Anything Else?

- To generate API documentation for spec.yaml run `redocly preview-docs spec.yaml ` 
- Facing TLS difficulties on MacOS with observatorium API [check this out ](https://github.com/observatorium/api/issues/192)

### CODE REVIEW

File: client/responses/responses.yaml [hunk 1]
```diff
@@ -121,3 +126,52 @@ components:
...
+          - streams
+        properties:
+          error:
+            type: string
+          errorType:
+            type: string
+          warnings:
+            type: array
+            items:
+              type: string
```
Comment (periklis) [inline client/responses/responses.yaml:142]: Can you provide a link/reference to validate these properties? I can't find anything in the protobuf or golang code.

File: client/spec.yaml [hunk 2]
```diff
@@ -231,4 +231,199 @@ paths:
...
+      operationId: getLogInstantQuery
+      description: |
+        You can evaluate instant queries for a tenant
+      responses:
+        "2XX":
+          description: instance query result for a tenant
+          content:
+            application/json:
+              schema:
+                $ref: ./responses/responses.yaml#/components/schemas/QueryResponse
```
Comment (periklis) [inline client/spec.yaml:393]: Shouldn't this be `InstantQueryResponse` here?

**Review by periklis — CHANGES_REQUESTED** [review]
In general it looks so far in a good shape. I added a couple of comments to correct wrong links and typos here and there. Besides these my only concern is if re-using the responses (e.g. `QueryResponse` or `InstantQueryResponse`) is the correct approach here. Maybe we should consider keeping the response objects for logs separately from metrics (cc @saswatamcode WDYT?)

File: client/spec.yaml [hunk 3]
```diff
@@ -231,4 +231,199 @@ paths:
...
+        - pushv1
+      summary: sends log entries to Loki
+      operationId: postlogEntries
+      requestBody:
+        description: You can send all the log entries of clusters to Loki
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: ./responses/responses.yaml#/components/schemas/LogWriteResponse
```
[the remark this answers is not in the document]
Comment (nyza99) [reply client/spec.yaml:249]: Request object is same as response object so, I am just reusing it.

File: client/spec.yaml [hunk 4]
```diff
@@ -231,4 +231,199 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                 $ref: ./responses/responses.yaml#/components/schemas/LogWriteResponse
+        "401":
+          description: error finding tenant ID
+        "403":
+          description: unauthorized
+        "500":
+          description: server side error while listing rules
```
[the remark this answers is not in the document]
Comment (nyza99) [reply client/spec.yaml:262]: yes fixed it!

File: client/spec.yaml [hunk 5]
```diff
@@ -231,4 +231,199 @@ paths:
...
+      - $ref: ./parameters/parameters.yaml#/components/parameters/tenant
+      - $ref: ./parameters/parameters.yaml#/components/parameters/query
+      - $ref: ./parameters/parameters.yaml#/components/parameters/startTS
+      - $ref: ./parameters/parameters.yaml#/components/parameters/endTS
+      - $ref: ./parameters/parameters.yaml#/components/parameters/limit
+      - in: query
+        name: step
+        description: Query resolution step width
+        schema:
+          type: string
```
[the remark this answers is not in the document]
Comment (nyza99) [reply client/spec.yaml:350]: Added!

File: client/spec.yaml [hunk 6]
```diff
@@ -231,4 +231,199 @@ paths:
...
+              schema:
+                $ref: ./responses/responses.yaml#/components/schemas/SeriesResponse
+        "4XX":
+          description: Bad request
+        "5XX":
+          description: Server side error
+  /api/logs/v1/{tenant}/loki/api/v1/query_range:
+    parameters:
+      - $ref: ./parameters/parameters.yaml#/components/parameters/tenant
+      - $ref: ./parameters/parameters.yaml#/components/parameters/query
```
[the remark this answers is not in the document]
Comment (nyza99) [reply client/spec.yaml:342]: done

File: client/spec.yaml (hunk 2 above)
Comment (nyza99) [reply client/spec.yaml:393]: It refers to InstantQueryResponse in responses

**Comment by periklis** [comment]
@nyza99 Can you push your commits addressing the review comments above which you see "fixed"?

File: client/responses/responses.yaml (hunk 1 above)
Comment (nyza99) [reply client/responses/responses.yaml:248]: > Can you provide a link/reference to validate these properties? I can't find anything in the protobuf or golang code.

I carried it from metric properties

File: client/spec.yaml [hunk 7]
```diff
@@ -227,8 +227,257 @@ paths:
...
+      operationId: postlogEntries
+      requestBody:
+        description: You can send all the log entries of clusters to Loki
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: ./models/models.yaml#/components/schemas/PushLogs
+      responses:
+        "201":
```
Comment (periklis) [inline client/spec.yaml:251]: This should be a 204.

**Review by periklis — COMMENTED** [review]
Looks good to me, besides one small issue

**Review by periklis — COMMENTED** [review]
Yippie!!!

**Comment by periklis** [comment]
Needs a rebase plus client code regeneration and imho we are set to go.

**Review by periklis — COMMENTED** [review]
Cool stuff!

**Comment by nyza99** [comment]
> Cool stuff!

**Comment by periklis** [comment]
@saswatamcode Any opinion on how to use the tags correctly to make a visual distinction for paths for logs and metrics. Looking at this feels like we have two labelsv1 paths etc. One needs to look on the right panel to see if they use the metrics or the logs endpoints
image
 
Should we add a prefix to the tags likes `metrics/labelsv1` and `logs/labelsv1`? Do you have any better option?

cc @nyza99

**Comment by saswatamcode** [comment]
Ah yes. I was wondering if this would happen. I think adding `metrics` and `logs` prefixes to the tags would make it distinct! 🙂

**Review by periklis — APPROVED** [review]

**Pull request merged**